### ISSUE TITLE
Store page boundaries and keys in infinite queries

### ISSUE THREAD

**Issue by clragon** [D-James-GH/cached_query#53]
### Is your feature request related to a problem? Please describe.
Currently, `InfiniteQueryState` holds a `List<T>`, which makes it impossible to determine which items belong to which page and to track their corresponding keys. We can therefore no longer distinguish pages.

### Describe the solution you'd like
I propose changing the `InfiniteQueryState` to hold a `List<List<T>>` instead of a `List<T>` to better organize items by page. Additionally, `InfiniteQueryState` should maintain a `List<Arg>` to track the keys associated with each page.

This approach would allow us to know both:
- Which items belong to which page.
- Which items correspond to which page key.

This is very similar to how React Query handles infinite queries. While using a `Map` might seem like an obvious solution, since `LinkedHashmap` is ordered, presume we are using bidirectional pagination, inserting a new page at the start would become difficult, especially since we cannot know the sorting order of arbitrary keys.
Using two arrays where indexes associate items is therefore more manageable.

### Additional context
A practical use case for this is managing headers or scroll bar page indicators, where it’s necessary to know the boundaries of each page and their respective keys.

This is a breaking change. However, it would be possible, and I would recommend this too, to implement a Getter on the `InfiniteQueryState` that returns a flattened version of the pages array. This could make this an unbreaking change, or at minimum make migration easier.

**Comment by D-James-GH** [comment]
I think the datatype of `InfiniteQueryState` should remain the same (`List<T>`). Currently each item in the list is a "page". Each page doesn't necessarily have to be a list (eg multiple posts). Imagine a situation where a page could be single item, like a page of a document/pdf. It's up to the user of the package to pass in the correct type of the page (which is usually a `List<SomeModel>`.

I don't see a problem with storing the result of `getNextArg` along with the page, this could be a useful enhancement.

**Comment by clragon** [comment]
Youre right that is a much better idea, and it prompted me to double check!

```ts
interface InfiniteData<TData, TPageParam = unknown> {
    pages: Array<TData>;
    pageParams: Array<TPageParam>;
}
```

React Query does in fact not opinionate whether the Result is a nested list, this is simply what we get when we define it as such. 
I think that would be a fine implementation.

**Comment by clragon** [comment]
I have tried out the pre-release of v3.0.0 and this works very well for what I want.
However, one thing I noticed is: `pageParams` is `Arg?` instead of `Arg`:

```dart
  /// The pageParams used to fetch the page of the same index
  List<Arg?>? get pageParams;
```

But it is unclear to me why this would be the case.
When `GetNextArg` returns null, this means there are no more pages - therefore, null should never be added to `pageParams`.
I would also not think to find null there ever. In my usage I had to explicitly convert the `pageParams` list to its non-null counterpart, because the rest of my code was not expecting a nullable list.

Is there a specific reason this is implemented in this way?

**Comment by clragon** [comment]
The `pageParams` being null as a whole in initial and loading states works fine for me!
`List<Arg>?` is a perfectly acceptable type for all my usecases. 
Just `List<Arg?>?` would be inconvenient.

**Comment by D-James-GH** [comment]
Added in `cached_query: ^3.0.0`
